client/web/src/components/externalServices/AddExternalServicesPage.tsx

```diff
@@ -54,30 +54,37 @@ export const AddExternalServicesPage: FC<AddExternalServicesPageProps> = ({
         setHasDismissedPrivacyWarning(true)
     }
 
-    const params = new URLSearchParams(search)
-    const id = params.get('id')
-    if (id) {
-        let externalService = allExternalServices[id]
-        if (externalService) {
-            if (externalService.kind === ExternalServiceKind.GITHUB) {
-                const appID = params.get('appID')
-                const installationID = params.get('installationID')
-                const baseURL = params.get('url')
-                const org = params.get('org')
-                if (appID && installationID && baseURL && org) {
-                    externalService = gitHubAppConfig(baseURL, appID, installationID, org)
+    const externalService = useMemo(() => {
+        const params = new URLSearchParams(search)
+        const id = params.get('id')
+        if (id) {
+            let externalService = allExternalServices[id]
+            if (externalService) {
+                if (externalService.kind === ExternalServiceKind.GITHUB) {
```

Comment:
This could be shortened with the elvis operator, I think

```suggestion
            if (externalService?.kind === ExternalServiceKind.GITHUB) {
```

Reply:
Not really, since we do not want to return the `AddExternalServicePage` on line 71 if the `externalService` is not defined.

Reply:
I also think the change can be made, because effectively you have this, right?
```js
if (externalService && externalService.kind === ExternalServiceKind.GITHUB) {
  // doing stuff
}
```

and you also have no `else` clauses anywhere in this function, so I guess it should be fine.

Reply:
Nope. It still won't work, see my previous comment.

Reply:
@kopancek 
Just to get this settled in my head :)

### We have this code
```js
    const externalService = useMemo(() => {
        const params = new URLSearchParams(search)
        const id = params.get('id')
        if (id) {
            let externalService = allExternalServices[id]
            if (externalService) {
                if (externalService.kind === ExternalServiceKind.GITHUB) {
                    const appID = params.get('appID')
                    const installationID = params.get('installationID')
                    const baseURL = params.get('url')
                    const org = params.get('org')
                    if (externalService === codeHostExternalServices.ghapp) {
                        externalService = gitHubAppConfig(baseURL, appID, installationID, org)
                    }
                }
            }
            return externalService
        }
        return null
    }, [search, codeHostExternalServices.ghapp])

    if (externalService) {
        return (
            <AddExternalServicePage
                telemetryService={telemetryService}
                externalService={externalService}
                autoFocusForm={autoFocusForm}
                externalServicesFromFile={externalServicesFromFile}
                allowEditExternalServicesWithFile={allowEditExternalServicesWithFile}
            />
        )
    }
```

### We have this desired behaviour
`we do not want to return the AddExternalServicePage on line 71 if the externalService is not defined.`

### My understanding
`we do not want to return the AddExternalServicePage` means that we should set `externalService` to `null` in this function so that we fail on `if (externalService)` check and don't create an `AddExternalServicePage` component.

### When do we return `null`
I can see that we return `null` explicitly in case of not having an `id` query parameter in the URL, but **if we have it**, then we return `null` only when `allExternalServices[id]` returns null. Otherwise, whether or not the external service is GitHub, we will return the non-null value.

### The final
Can you pls show me why changing the 
```js
if (externalService) {
  if (externalService.kind === ExternalServiceKind.GITHUB)
```

to
```js
if (externalService && externalService.kind === ExternalServiceKind.GITHUB)
```

or
```js
if (externalService?.kind === ExternalServiceKind.GITHUB)
```

should break anything in terms of `externalService` nullability, because it remains unchanged if we enter this if.

I've read the code multiple times, but I'm clearly missing something here, sorry 😅

Reply:
OK, sorry for wasting everybody's time here. I was looking at the wrong branch locally when making my comments 🤦 This should be fixed now 👍